Additional context and questions for operator capabilities (#2680)
operator-framework/operator-sdk: doc/operator-capabilities.md

## Level 1 - Basic Install

> Automated application provisioning and configuration management. This first capability level means your operator can fully provision an application through a custom resource, allowing all installation configuration details to be specified in the CR. It should also be possible to install the operator itself in multiple ways (kubectl, OLM, Catalog source). Any configuration required to make the Operand run should be configured through the CR if possible. Avoid the practice of requiring the user to create/manage configuration files outside of Kubernetes/OpenShift.

Reviewer: ```suggestion
Automated application provisioning and configuration management. This first capability level means your operator can fully provision an application through a custom resource, allowing all installation configuration details to be specified in the CR. It should also be possible to install the operator itself in multiple ways (kubectl, OLM, Catalog source). Any configuration required to make the Operand run should be configured through the CR if possible. Avoid the practice of requiring the user to create/manage configuration files outside of Kubernetes.
```

Reviewer: Removed OpenShift from there. OpenShift is Kubernetes.

3. Can you set operand configuration in the CR? If so, what configuration is supported for each operand?

> 4. Can you override the operand images through the CR or an env var?

Reviewer: ```suggestion
4. Can you override the operand images through the CR or an environment variable of the Operator deployment?
```

5. Does the managed application / workload get updated in a non-disruptive fashion when the configuration of the CR is changed?

6. Does the status of the CR reflect that configuration changes are currently applied?

7. What additional operand configuration could still be added?

8. Do all of the instantiated CRs include a status block? If so, does it provide enough insight to the user about the application state?

9. Do all of your CRs have documentation listing valid values and mandatory fields?

> 10. If your operator includes a CSV, are all images used listed in the CSV under relatedImages?

Reviewer: ```suggestion
10. If your operator is packaged for OLM, does its CSV list all images used in the CSV under `spec.relatedImages`?
```

---

## Level 2 - Seamless Upgrades

> Seamless upgrades mean the upgrade is as easy as possible for the user. You should support seamless upgrades of both your operator and operand, these would normally go hand in hand, an upgrade of the operator would automatically ensure the instantiated resources for each CR are in the new desired state and which would upgrade your operand. Upgrade may also be defined multiple ways, such as updating the software of the operand - and other internals specific to the application - such as schema migrations. It should be very clear what is upgraded when this takes place, and what is not.

Reviewer: The upgrade of the Operator and Operand don't necessarily go hand-in-hand though. This is just one option. Many Operators control the Operand version based on a property in the CR. The `Seamless Upgrade` capability would recommend that the Operator is capable of upgrading the Operand either in this manner or automatically to the version the Operator prefers.

Author: @dmesser I agree with you, not all operators have a 1-1 relationship between operator and operand. I am under the impression this is still a best practice though if it is possible to achieve with your operator and operand.

This paragraph does call out that this is the normal case but there are other defined upgrades also.

## Level 3 - Full Lifecycle

> It should be possible to backup and restore the operand from the operator itself without any additional manual intervention other than triggering these operations. The operand data that should be backed up is any stateful data managed by the operand. You don’t need to backup the CR itself. You do not need to backup the k8s resources created by the operator as the operator should return all resources to the same state if the CR is recreated. If your operator does not already setup the operand with other k8s resilient best practices, this should be completed to achieve this capability level. This includes liveness and readiness probes, multiple replicas, rolling deployment strategies, pod disruption budgets, CPU and memory requests and limits.

Reviewer: I think we can omit mentioning the backup of the CR but instead advise that the Operator recreates CR as part of the restore from a backup.

Author: Thanks @dmesser , I have reworded this a small bit now. I am under the impression that restore would be any operand stateful data. The created CR should be stored somewhere else and re-instantiated by the user, would you agree with this?

5. Does your Operator allow adding/removing read-only slave instances of your Operator?

6. Does your operand have a Liveness probe?

7. Does your operand have a Readiness probe which will fail if any aspect of the operand is not ready? e.g. if the connection to the database fails.

8. Does your operand use a rolling deployment strategy?

9. Does your operator create a PodDsruptionBudget resource for your operand pods?

10. Does your operand have CPU requests and limits set?

---

## Level 4 - Deep Insights

> Setup full monitoring and alerting for your operand. All resources such as Prometheus rules (alerts) and Grafana dashboards should be created by the operator when the operand CR is instantiated. The RED method is a good place to start with knowing what metrics to expose.

Reviewer: Provide reference to the RED method.

Author: Any preference on which link to include here? 
https://www.weave.works/blog/the-red-method-key-metrics-for-microservices-architecture/ or https://grafana.com/blog/2018/08/02/the-red-method-how-to-instrument-your-services/

Reviewer: I think we would prefer an proper definition somewhere rather than a blog article from practitioners.

Author: Thanks @dmesser . I am unable to find any reference outside of these. I can add in a short description here or remove the sentence.

Reviewer: Yeah, why not define it in a foot note. If not possible in a concise fashion, probably better to remove it.

Author: Updated now @dmesser . Let me know what you think.

2. Does your Operator expose Operand alerts?

> 4. Do you have SOPs for each alert?

Reviewer: Define SOP.

5. Does you operator create critical alerts when the service is down and warning alerts for all other alerts?

6. Does your Operator watch the Operand to create alerts?

7. Does your Operator emit custom Kubernetes events?

8. Does your Operator expose Operand performance metrics?

---

## Level 5 - Auto Pilot

> The final piece of the puzzle in eliminating/significantly reducing  any remaining manual intervention in managing the operand. The operator should automatically pick up on load of the operand and auto scale based on thresholds. The operator should automatically fix an unhealthy operand. The operator should tune the operands performance, this could include changing the node the pods are running on or modifying operand configuration.

Reviewer: ```suggestion
The highes capability level aims to significantly reduce/eliminate any remaining manual intervention in managing the operand. The operator should configure the Operand to auto-scale as load picks up. The Operator should understand  the application-level performance indicators and determine when it's healthy and performing well. The operator should attempt automatically fix an unhealthy operand. The operator should tune the operands performance, this could include scheduling on another node the pods are running on or modifying operand configuration.
```
